src/uucore/src/lib/features/encoding.rs

```diff
@@ -42,138 +30,115 @@ pub enum Format {
-            },
-        }
-    }
+        output.extend(string.as_bytes());
 
-    #[must_use]
-    pub fn line_wrap(mut self, wrap: usize) -> Self {
-        self.line_wrap = wrap;
-        self
+        Ok(())
     }
+}
 
-    #[must_use]
-    pub fn ignore_garbage(mut self, ignore: bool) -> Self {
-        self.ignore_garbage = ignore;
-        self
-    }
+impl SupportsFastEncode for Encoding {
+    // Adapted from `encode_append` in the "data-encoding" crate
+    fn encode_to_vec_deque(&self, input: &[u8], output: &mut VecDeque<u8>) -> UResult<()> {
+        let output_len = output.len();
 
-    pub fn decode(&mut self) -> DecodeResult {
-        let mut buf = vec![];
-        self.input.read_to_end(&mut buf)?;
-        if self.ignore_garbage {
-            buf.retain(|c| self.alphabet.contains(c));
-        } else {
-            buf.retain(|&c| c != b'\r' && c != b'\n');
-        };
-        decode(self.format, &buf)
-    }
+        output.resize(output_len + self.encode_len(input.len()), 0_u8);
 
-    pub fn encode(&mut self) -> Result<String, EncodeError> {
-        let mut buf: Vec<u8> = vec![];
-        match self.input.read_to_end(&mut buf) {
-            Ok(_) => encode(self.format, buf.as_slice()),
-            Err(_) => Err(EncodeError::InvalidInput),
-        }
+        let make_contiguous_result = output.make_contiguous();
+
+        self.encode_mut(input, &mut (make_contiguous_result[output_len..]));
+
+        Ok(())
     }
 }
 
-// NOTE: this will likely be phased out at some point
-pub fn wrap_print<R: Read>(data: &Data<R>, res: &str) {
-    let stdout = io::stdout();
-    wrap_write(stdout.lock(), data.line_wrap, res).unwrap();
+pub trait SupportsFastDecode {
+    fn decode_into_vec(&self, input: &[u8], output: &mut Vec<u8>) -> UResult<()>;
 }
 
-pub fn wrap_write<W: Write>(mut writer: W, line_wrap: usize, res: &str) -> io::Result<()> {
-    use std::cmp::min;
+impl SupportsFastDecode for ZEightFiveWrapper {
+    fn decode_into_vec(&self, input: &[u8], output: &mut Vec<u8>) -> UResult<()> {
+        if input.first() == Some(&b'#') {
+            return Err(USimpleError::new(1_i32, "error: invalid input".to_owned()));
+        }
 
-    if line_wrap == 0 {
-        return write!(writer, "{res}");
-    }
+        // According to the spec we should not accept inputs whose len is not a multiple of 4.
+        // However, the z85 crate implements a padded encoding and accepts such inputs. We have to manually check for them.
+        if input.len() % 4_usize != 0_usize {
+            return Err(USimpleError::new(
+                1_i32,
+                "error: invalid input (length must be multiple of 4 characters)".to_owned(),
+            ));
+        };
 
-    let mut start = 0;
-    while start < res.len() {
-        let end = min(start + line_wrap, res.len());
-        writeln!(writer, "{}", &res[start..end])?;
-        start = end;
+        let decode_result = match z85::decode(input) {
+            Ok(ve) => ve,
```

Comment:
Codecov says that this (and many other) lines weren't tested. This is surprising. Is this a false negative, or is there really no test for many lines of code in this PR?

Reply:
I'm not exactly sure how this logic would be unit tested, or how beneficial that would be compared to other approaches. I do agree that it would be nice to have more assurance that these new implementations are correct.

The tests in `test_base32.rs`, `test_base64.rs`, and `test_basenc.rs` are covering basic uses with small inputs, but because of the new buffering logic, I think fuzz or property testing, which could generate large random inputs, would be a good way to make sure that there are no subtle edge cases.